File: __tests__/lib/byok/chat.test.ts
/**
 * @jest-environment node
 */
import { buildChatBody, extractDelta, readStream } from '@/lib/byok/chat'

describe('buildChatBody', () => {
  const msgs = [
    { role: 'user' as const, content: 'what do I spend on food?' },
    { role: 'assistant' as const, content: '45 KWD last month.' },
  ]

  it('builds the Anthropic messages format (system as top-level string)', () => {
    const body = buildChatBody({ id: 'anthropic' }, 'You are Fajrak.', msgs, 'claude-x', true) as any
    expect(body.model).toBe('claude-x')
    expect(body.system).toBe('You are Fajrak.')
    expect(body.max_tokens).toBe(1024)
    expect(body.stream).toBe(true)
    expect(body.messages).toEqual(msgs.map(m => ({ role: m.role, content: m.content })))
  })

  it('builds Gemini contents with assistant mapped to role model', () => {
    const body = buildChatBody({ id: 'gemini' }, 'Sys', msgs, 'gemini-2.5', false) as any
    expect(body.system_instruction).toEqual({ parts: [{ text: 'Sys' }] })
    expect(body.contents).toEqual([
      { role: 'user', parts: [{ text: 'what do I spend on food?' }] },
      { role: 'model', parts: [{ text: '45 KWD last month.' }] },
    ])
    // Gemini signals streaming via the streamGenerateContent URL, not a body flag.
    expect(body.stream).toBeUndefined()
  })

  it('builds OpenAI-compatible bodies with a prepended system message for every proxy provider', () => {
    for (const id of ['openai', 'nvidia-nim', 'openrouter', 'ollama']) {
      const body = buildChatBody({ id }, 'Sys', msgs, 'm', true) as any
      expect(body.temperature).toBe(0.2)
      expect(body.stream).toBe(true)
      expect(body.messages).toEqual([
        { role: 'system', content: 'Sys' },
        ...msgs,
      ])
    }
  })
})

describe('extractDelta', () => {
  it('extracts Anthropic content_block_delta text only', () => {
    expect(
      extractDelta('anthropic', { type: 'content_block_delta', delta: { text: 'Hel' } })
    ).toBe('Hel')
    expect(extractDelta('anthropic', { type: 'message_start' })).toBe('')
    expect(extractDelta('anthropic', { type: 'content_block_delta', delta: {} })).toBe('')
  })

  it('extracts the first Gemini candidate part text', () => {
    expect(
      extractDelta('gemini', {
        candidates: [{ content: { parts: [{ text: 'world' }] } }],
      })
    ).toBe('world')
    expect(extractDelta('gemini', { candidates: [] })).toBe('')
    expect(extractDelta('gemini', {})).toBe('')
  })

  it('extracts OpenAI delta text (with message fallback)', () => {
    expect(extractDelta('openrouter', { choices: [{ delta: { content: 'ab' } }] })).toBe('ab')
    expect(extractDelta('ollama', { choices: [{ message: { content: 'cd' } }] })).toBe('cd')
    expect(extractDelta('openai', { choices: [{ delta: {} }] })).toBe('')
    expect(extractDelta('nvidia-nim', {})).toBe('')
  })
})

describe('readStream', () => {
  function sseResponse(chunks: string[]): Response {
    const stream = new ReadableStream<Uint8Array>({
      start(controller) {
        for (const c of chunks) controller.enqueue(new TextEncoder().encode(c))
        controller.close()
      },
    })
    return new Response(stream, { status: 200 })
  }

  it('collects incremental Delta text and skips non-text frames', async () => {
    const body = [
      'data: {"type":"content_block_delta","delta":{"text":"Hello"}}\n\n',
      'data: {"type":"content_block_delta","delta":{"text":" world"}}\n\n',
      'data: {"type":"message_delta","delta":{"stop_reason":"end_turn"}}\n\n',
      'event: ping\ndata: [DONE]\n\n',
    ].join('')

    const seen: string[] = []
    await readStream(sseResponse([body]), 'anthropic', t => seen.push(t))
    expect(seen).toEqual(['Hello', ' world'])
  })

  it('rejoins frames split across network chunks', async () => {
    const seen: string[] = []
    await readStream(
      sseResponse(['data: {"type":"content_block_delta","delta":{"text":"He', 'llo"}}\n\n']),
      'anthropic',
      t => seen.push(t)
    )
    expect(seen).toEqual(['Hello'])
  })

  it('ignores malformed frames instead of throwing', async () => {
    const seen: string[] = []
    await readStream(
      sseResponse(['data: not-json\n\ndata: {"choices":[{"delta":{"content":"ok"}}]}\n\n']),
      'openai',
      t => seen.push(t)
    )
    expect(seen).toEqual(['ok'])
  })

  it('skips comment/event lines and blank data', async () => {
    const seen: string[] = []
    await readStream(
      sseResponse([': keep-alive\nevent: ping\ndata:\n\ndata: [DONE]\n\n']),
      'openai',
      t => seen.push(t)
    )
    expect(seen).toEqual([])
  })

  it('throws when the response has no body', async () => {
    await expect(readStream(new Response(null, { status: 200 }), 'openai', () => {})).rejects.toThrow(
      'No response body'
    )
  })
})

File: mobile/fajrak_flutter/test/services/byok/chat_test.dart
import 'dart:async';
import 'dart:convert';

import 'package:fajrak/services/byok/chat.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;

Map<String, dynamic> _decode(Object? body) =>
    body is Map<String, dynamic>
        ? body
        : jsonDecode(body as String) as Map<String, dynamic>;

void main() {
  final system = 'You are a helpful assistant.';
  final msgs = [
    ChatMsg(role: 'user', content: 'Hello'),
    ChatMsg(role: 'assistant', content: 'Hi there'),
  ];

  group('buildChatBody', () {
    test('openai shape (openai / nvidia-nim / openrouter / ollama)', () {
      final body = _decode(buildChatBody('openai', system, msgs, 'gpt-5.4-mini', true));
      expect(body['model'], 'gpt-5.4-mini');
      expect(body['temperature'], 0.2);
      expect(body['stream'], true);
      expect(body['messages'], [
        {'role': 'system', 'content': system},
        {'role': 'user', 'content': 'Hello'},
        {'role': 'assistant', 'content': 'Hi there'},
      ]);
    });

    test('anthropic shape includes system, max_tokens, no merged system message', () {
      final body = _decode(buildChatBody('anthropic', system, msgs, 'claude-sonnet-4-6', false));
      expect(body['model'], 'claude-sonnet-4-6');
      expect(body['system'], system);
      expect(body['max_tokens'], 1024);
      expect(body['temperature'], 0.2);
      expect(body['stream'], false);
      expect(body['messages'], [
        {'role': 'user', 'content': 'Hello'},
        {'role': 'assistant', 'content': 'Hi there'},
      ]);
    });

    test('gemini shape maps assistant->model and wraps parts', () {
      final body = _decode(buildChatBody('gemini', system, msgs, 'gemini-2.5-pro', true));
      expect(body['model'], 'gemini-2.5-pro');
      expect(body['system_instruction'], {'parts': [{'text': system}]});
      expect(body['generationConfig'], {'temperature': 0.2});
      expect(body['contents'], [
        {'role': 'user', 'parts': [{'text': 'Hello'}]},
        {'role': 'model', 'parts': [{'text': 'Hi there'}]},
      ]);
    });
  });

  group('extractDelta', () {
    test('anthropic content_block_delta -> text', () {
      final d = extractDelta('anthropic', {
        'type': 'content_block_delta',
        'delta': {'text': 'Hello'},
      });
      expect(d, 'Hello');
    });

    test('anthropic non-delta frame -> empty', () {
      expect(extractDelta('anthropic', {'type': 'message_start'}), '');
    });

    test('gemini -> first candidate first part text', () {
      final d = extractDelta('gemini', {
        'candidates': [
          {'content': {'parts': [{'text': 'World'}]}},
        ],
      });
      expect(d, 'World');
    });

    test('openai-compatible chooses delta then message', () {
      final viaDelta = extractDelta('openai', {
        'choices': [{'delta': {'content': 'A'}}],
      });
      expect(viaDelta, 'A');

      final viaMessage = extractDelta('nvidia-nim', {
        'choices': [{'message': {'content': 'B'}}],
      });
      expect(viaMessage, 'B');
    });

    test('missing fragments -> empty string', () {
      expect(extractDelta('openai', {'choices': []}), '');
      expect(extractDelta('gemini', {'candidates': []}), '');
      expect(extractDelta('anthropic', {'type': 'content_block_delta'}), '');
    });
  });

  group('readStream', () {
    final openaiFrame =
        'data: ${jsonEncode({'choices': [{'delta': {'content': 'tok'}}]})}\n\n';
    final done = 'data: [DONE]\n\n';

    test('accumulates deltas across chunk-split boundaries', () async {
      final body = openaiFrame + done;
      final streamed = http.StreamedResponse(
        Stream.fromIterable([utf8.encode(body)]),
        200,
      );

      final out = <String>[];
      await readStream(streamed, 'openai', out.add);
      expect(out, ['tok']);
    });

    test('tolerates malformed frames and splits mid-frame', () async {
      // Two "tok" frames interleaved with a malformed frame, split mid-line.
      final a = 'data: {"choices":[{"delta":{"content":"t';
      final b = 'o"}}]}\n\njunk\n\ndata: [DONE]';
      final streamed = http.StreamedResponse(
        Stream.fromIterable([
          utf8.encode(a),
          utf8.encode(b),
        ]),
        200,
      );

      final out = <String>[];
      await readStream(streamed, 'openai', out.add);
      expect(out, ['to']);
    });

    test('silent-stall (S1): >20s gap throws a TimeoutException', () async {
      final streamed = http.StreamedResponse(
        // First event immediately, then a 25s gap, then a late event.
        Stream.periodic(const Duration(seconds: 25), (_) => utf8.encode('x')),
        200,
      );
      final out = <String>[];

      await expectLater(
        readStream(
          streamed,
          'openai',
          out.add,
          stallTimeout: const Duration(milliseconds: 50),
        ),
        throwsA(isA<TimeoutException>()),
      );
    });

    test('stop flag swallows a racing TimeoutException (S1)', () async {
      final streamed = http.StreamedResponse(
        Stream.periodic(const Duration(seconds: 25), (_) => utf8.encode('x')),
        200,
      );
      var stopped = false;
      // Stop fires before the stall timeout: the timeout must be swallowed.
      Timer(const Duration(milliseconds: 10), () => stopped = true);
      final out = <String>[];

      await expectLater(
        readStream(
          streamed,
          'openai',
          out.add,
          isStopped: () => stopped,
          stallTimeout: const Duration(milliseconds: 200),
        ),
        completes,
      );
    });
  });
}
